### PR TITLE
Make benchmark suite check for success and correct output

### PR DESCRIPTION
**Pull request by quantum5**
`test_zzz_java_nbody` is broken and not easily fixable. It's a non-core test.

### CODE REVIEW

File: tests/test_benchmark.py [hunk 1]
```diff
@@ -978,12 +981,12 @@ def test_zzz_poppler(self):
...
-          Module.arguments = ['-?'];
-          Module.printErr = function(){};
+          Module['arguments'] = ['-?'];
+          Module['printErr'] = function(){};
         } else {
           // Add 'filename' after 'input.pdf' to write the output so it can be verified.
-          Module.arguments = ['-scale-to', '1024', 'input.pdf',  '-f', '1', '-l', '' + benchmarkArgumentToPageCount[benchmarkArgument]];
-          Module.postRun = function() {
+          Module['arguments'] = ['-scale-to', '1024', 'input.pdf',  '-f', '1', '-l', '' + benchmarkArgumentToPageCount[benchmarkArgument]];
+          Module['postRun'] = function() {
```
Comment (sbc100) [inline tests/test_benchmark.py]: Why this change?

Comment (quantum5) [reply tests/test_benchmark.py]: Because closure mangles `postRun` into something else, and it never gets run. Since we never bothered checking the output, the benchmark passed anyways.

File: tests/test_benchmark.py [hunk 2]
```diff
@@ -52,12 +52,15 @@ class Benchmarker(object):
...
 
-  def bench(self, args, output_parser=None, reps=TEST_REPS):
+  def bench(self, args, output_parser=None, reps=TEST_REPS, expected_output=None):
     self.times = []
     self.reps = reps
     for i in range(reps):
       start = time.time()
       output = self.run(args)
+      if expected_output is not None and expected_output not in output:
+        raise ValueError('Incorrect benchmark output:\n' + output)
```
Comment (kripken) [inline tests/test_benchmark.py]: can use `self.assertContained` like the main test suite, it will show a nice diff on errors.

File: tests/test_benchmark.py [hunk 3]
```diff
@@ -354,7 +357,7 @@ def setUpClass(self):
 
   # avoid depending on argument reception from the commandline
   def hardcode_arguments(self, code):
-    if not code:
+    if not code or 'int main()' in code:
```
Comment (kripken) [inline tests/test_benchmark.py]: this looks bad - we need to hardcode that argument. where is it not possible to do so?

File: tests/test_benchmark.py (hunk 2 above)
Comment (quantum5) [reply tests/test_benchmark.py]: I can't use `assertContained` because this is not in a test. This is in the benchmarker class.

File: tests/test_benchmark.py (hunk 3 above)
Comment (quantum5) [reply tests/test_benchmark.py]: Do we need to hardcode that argument? If the source code uses `int main()`, it doesn't care what the arguments are.

File: tests/test_benchmark.py (hunk 2 above)
Comment (kripken) [reply tests/test_benchmark.py]: Oh, ok.

File: tests/test_benchmark.py (hunk 3 above)
Comment (kripken) [reply tests/test_benchmark.py]: It would be nice if all benchmarks followed that convention, but yeah, I guess some of the non-core ones don't?
